### ISSUE TITLE
Reserve threads for metadata and status operations

### ISSUE THREAD

**Issue by adriansev** [xrootd/xrootd#927]
Hi! It would be fantastic if for metadata and/or status queries some threads could be reserved 
(of course configurable in xrd.sched).   
When the xrootd server is under load and maximum number of threads are reached one cannot query about the status of the server of file metadata (operations with ~ 0 latency and 0 resource consumptions.).   
The increasing of maxt would not help as all of them will be used until the server is bogged down with transfers.

**Comment by jmuf** [comment]
(remote "metadata and status" might still clog up authentication, so perhaps not 0 resource? but perhaps per-authenticationtype threadpools would at least allow unauthenticated localhost queries to get through?)

**Comment by adriansev** [comment]
well, 0 resource i meant that thread allocation time is on the order of ms to s in contrast to thousands (or maxt number) of transfers that block any other query to xrootd server. so, even with auth it is (should be) ok if the operations are of the status type `xrdfs query config` and `xrdfs query stats` and mostly ok for metadata type operations : `xrdfs stat`, `xrdfs statvfs`, `xrdfs query space`, `xrdfs query xattr` and `xrdfs ls`)

**Comment by abh3** [comment]
Before we start re-architecting the whole insides of XRootD, be aware that the model we use is run-to-completion which is usually the best one in the type of workflows we see most often. That means that thread pools provide limited improvement (actually almost none) this is because we need a thread to just service the socket and once we have it doesn’t matter what we do with it. In this kind of model thread priority would be far more effective. Unfortunately, Linux doesn’t seem to do a good job honoring the priority (well it didn’t some time ago so we don’t bother using it). In general, if you have a server that is bogged down then the way to properly solve this is to bring up another server to balance out the load.

Andy

**Comment by adriansev** [comment]
Well, i do not know the insides of xrootd, i just assumed that you have a threadpool (given the format of xrd.sched) .. also i do not know how queries/requests are dispatched (and if for xrootd there is a difference between query and request) .. so, from your answer i get that with the current design is not possible to make the service responsive when is under heavy IO load... thank you for info!

**Comment by abh3** [comment]
Hi Adrian,

I didn’t quite say that. It is possible if Linux thread priorities actually did something reasonable. Perhaps they do but they were hardly reasonable a decade ago. So, there’s room for experimentation.

Andy

**Comment by adriansev** [comment]
Well, just a thought : would be possible for the query requests (or even for metadata) to set the thread to SCHED_FIFO? (i am assuming that right now the threads are SCHED_OTHER)   
I am not sure if you can determine what command create the socket and the corresponding thread but you could elevate the thread priority afterwards (maybe like in this [link](http://www.yonch.com/tech/82-linux-thread-priority)? Does it make sense?

**Comment by abh3** [comment]
Yes, it is a possibility. Not clear it will help much should a server be completely bogged down, but certainly worth exploring. I'll reopen this and list it as an enhancement,

**Comment by adriansev** [comment]
Thanks a lot!

**Comment by abh3** [comment]
I am closing this as essentially a duplicate of issue #1051  We can track it there which is a more recent post.
